ldms/python/ldmsd/parser_util.py

```diff
@@ -0,0 +1,1005 @@
+            return None
+        if type(config['plugins']) is not dict:
+            print(f'{LDMS_YAML_ERR}')
+            print(f'store {DICT_ERR}')
+            print(f'e.g. plugins:')
+            print(f'       meminfo1 :')
+            print(f'         name      : meminfo')
+            print(f'         interval  : 1.0s')
+            print(f'         config    : [ { schema : meminfo }, { ... : ... } ]')
+            sys.exit(22)
+        plugins = {}
+        plugn_spec = config['plugins']
+        for plugn in plugn_spec:
+            if plugn in plugins:
+                print(f'Duplicate plugin name "{plugin_name}". '
+                      f'Plugin must be unique within a group.')
+                sys.exit(22)
+            check_plugin_config(plugn, plugn_spec)
+            plugins[plugn] = plugn_spec[plugn]
+        return plugins
+
+    def __init__(self, client, name, cluster_config, args):
+        """
+        Build configuration groups out of the YAML configuration
+        """
+        self.client = client
+        self.name = name
+        self.args = args
+        self.cluster_config = cluster_config
+        self.advertisers = {}
+        self.prdcr_listeners = {}
+        self.daemons = self.build_daemons(cluster_config)
+        self.plugins = self.build_plugins(cluster_config)
+        self.aggregators = self.build_aggregators(cluster_config)
+        self.producers = self.build_producers(cluster_config)
+        self.updaters = self.build_updaters(cluster_config)
+        self.stores = self.build_stores(cluster_config)
+        self.samplers = self.build_samplers(cluster_config)
+
+    def ldmsd_arg_list(self, local_path, dmn_grp, dmn):
+        start_list = [ 'ldmsd' ]
+        for ep in self.daemons[dmn_grp][dmn]['endpoints']:
+            if self.daemons[dmn_grp][dmn]['endpoints'][ep]['maestro_comm'] is True:
+                ep_ = self.daemons[dmn_grp][dmn]['endpoints'][ep]
+                start_list.append('-x')
+                start_list.append(f'{ep_["xprt"]}:{ep_["port"]}')
+                auth = check_opt('auth', ep_)
+                if auth:
+                    auth_plugin = check_opt('plugin', ep_['auth'])
+                    auth_opt = check_opt('conf', ep_)
+                    start_list.append('-a')
+                    start_list.append(auth_plugin)
+                    if auth_opt:
+                        if len(auth_opt.split('=')) < 2:
+                            auth_opt = f'conf={auth_opt}'
+                        start_list.append('-A')
+                        start_list.append(auth_opt)
+        start_list.append('-c')
+        start_list.append(f'{local_path}/{dmn}.conf')
+        start_list.append('-r')
+        start_list.append(f'{local_path}/{dmn}.pid')
+        start_list.append('-l')
+        start_list.append(f'{local_path}/{dmn}.log')
+        start_list.append(f'-F')
+        return start_list
+
+    def write_advertisers(self, dstr, dmn_grp, dname, auth_list):
```

Comment:
A whoooole bunch of these methods take "dstr" as an option. However, as far as I can tell not one of them actually needs to read the contents of that dstr, except that they do "dstr +=". I think it would make a lot of sense not do that, because the "+=" can always be done in the caller.

So unless I am missing something, I would recommend we redo all of these to not take "dstr" as an parameter.

Reply:
Some methods return a tuple, rather than solely a string. This was done to remain consistent in how class methods are called/returned to avoid confusion, but I'm fine with going with your suggestion if that's what everyone wants.

Reply:
You can still do += after breaking apart the tuple. I'd really like to see it consistent in the other direction.

Sending the suggests to me, as a programmer, that the code is going to use/modify that string in some way. If it doesn't do that, then you need to explicitly document the contract you are making with the caller "I won't modify this, only append new data".

But the method doesn't take that as a parameter, you save yourself some documentation. It is impossible for the method to change something it doesn't have. Its just good programming practice to only pass a method the things it actually needs to accomplish its goal.